iceoryx_dust/container/include/iox/detail/fixed_position_container.inl

```diff
@@ -51,6 +55,166 @@ inline FixedPositionContainer<T, CAPACITY>::~FixedPositionContainer() noexcept
+}
+
+template <typename T, uint64_t CAPACITY>
+template <typename RhsType>
+inline void FixedPositionContainer<T, CAPACITY>::copy_and_move_impl(RhsType&& rhs) noexcept
+{
+    // we already make sure rhs is always passed by std::move() for move case, therefore
+    // the result of "decltype(rhs)" is same as "decltype(std::forward<RhsType>(rhs))"
+    static_assert(std::is_rvalue_reference<decltype(rhs)>::value
+                      || (std::is_lvalue_reference<decltype(rhs)>::value
+                          && std::is_const<std::remove_reference_t<decltype(rhs)>>::value),
+                  "RhsType must be const lvalue reference or rvalue reference");
+
+    constexpr bool is_move = std::is_rvalue_reference<decltype(rhs)>::value;
+
+    IndexType i{Index::FIRST};
+
+    // this can update member faster
+    IndexType next_used_cache{Index::INVALID};
+    IndexType next_free_cache{Index::INVALID};
+    bool overwrite_used_slot{false};
+    bool overwrite_free_slot{true};
+
+    auto rhs_it = (std::forward<RhsType>(rhs)).begin();
+
+
+    // transfer src data to destination
+    for (; rhs_it.to_index() != Index::INVALID; ++i, ++rhs_it)
+    {
+        if (m_status[i] == SlotStatus::USED)
+        {
+#if __cplusplus >= 201703L
+            if constexpr (is_move)
+            {
+                m_data[i] = std::move(*rhs_it);
+            }
+            else
+            {
+                m_data[i] = *rhs_it;
+            }
+#else
+            // We introduce a helper struct primarily due to the test case
+            // e1cc7c9f-c1b5-4047-811b-004302af5c00. It demands compile-time if-else branching
+            // for classes that are either non-copyable or non-moveable.
+            // Note: With C++17's 'if constexpr', the need for these helper structs (labeled "C++ 14 feature"
+            // in fixed_position_container.hpp) can be eliminated.
+            AssignmentHelper<is_move>::assign(m_data[i], MoveHelper<is_move>::move_or_copy(rhs_it));
+#endif
+            // update next used cache
+            next_used_cache = static_cast<IndexType>(m_next[i]);
+            overwrite_used_slot = true;
+        }
+        // use ctor to avoid UB for non-initialized free slots
+        else
+        {
+#if __cplusplus >= 201703L
+            if constexpr (is_move)
+            {
+                new (&m_data[i]) T(std::move(*rhs_it));
+            }
+            else
+            {
+                new (&m_data[i]) T(*rhs_it);
+            }
+#else
+            // Same as above
+            CtorHelper<is_move>::construct(m_data[i], MoveHelper<is_move>::move_or_copy(rhs_it));
+#endif
+
+            // update next free cache
+            next_free_cache = static_cast<IndexType>(m_next[i]);
+            overwrite_free_slot = true;
+        }
+
+        m_status[i] = SlotStatus::USED;
+        m_next[i] = static_cast<IndexType>(i + 1U);
+    }
+
+    // update the member only if rhs is not empty to ensure that erase() functions correctly.
```

Comment:
I think the following code can be massively simplified if you take inspiration from the `clear` method. The `erase` method is not required since the internal structure is rebuild anyway.

Reply:
See 9eff9ce.